### PR TITLE
290 enrich events with coordinates

### LINKED ISSUE #290: Enrich Events with coordinates
### As a
frontend developer

### I want
i want my incoming events to have coordiantes.

### Because
that way i can display them in the map view.

## Tasks
- [ ] Open Pull Request
- [ ] share your insights with the Team
- [ ] enrich notifications with coordinates
- [ ] create a map function that does that

## Acceptance Criteria
- [ ] Changelog adjusted
- [ ] Documented
- [ ] Merged in dev
- [ ] Linter is okay
- [ ] Tests are written and green
- [ ] Code Review done
- [ ] Reviewed by PO

### PR DESCRIPTION
**Pull request by Quacck**
_implements_ #290

### CODE REVIEW

**Review by ma-ro — APPROVED** [review]

File: core/src/main/java/org/bptlab/cepta/operators/EnrichDelayWithCoordinatesFunction.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import org.bptlab.cepta.utils.notification.NotificationHelper;
+import org.bson.Document;
+
+import java.util.Collections;
+import java.util.Hashtable;
+import java.util.List;
+
+
+public class EnrichDelayWithCoordinatesFunction extends RichFlatMapFunction<NotificationOuterClass.Notification, NotificationOuterClass.Notification> {
+    /* This Function takes an inputstream of DelayNotifications and enriches the events
```
Comment (skydivin4ng3l) [inline core/src/main/java/org/bptlab/cepta/operators/EnrichDelayWithCoordinatesFunction.java:24]: A stream of Notifications, if they have a delayNotification with stationId they get enriched

File: core/src/main/java/org/bptlab/cepta/operators/EnrichDelayWithCoordinatesFunction.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    private Hashtable<String, CoordinateOuterClass.Coordinate> coordinateMapping =  new Hashtable<>();
+
+    public Hashtable<String, CoordinateOuterClass.Coordinate> getMapping(){ return coordinateMapping;}
+
+    private MongoConfig mongoConfig = new MongoConfig();
+    private transient MongoClient mongoClient;
+    private String tableName;
+
+    private boolean readInStationData(){
+        MongoDatabase database = mongoClient.getDatabase("replay"/*mongoConfig.getName()*/);
```
Comment (skydivin4ng3l) [inline core/src/main/java/org/bptlab/cepta/operators/EnrichDelayWithCoordinatesFunction.java]: maybe we should also make the database a member variable? to be able to specifiy which db hosts the "base data"

File: core/src/main/java/org/bptlab/cepta/operators/EnrichDelayWithCoordinatesFunction.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+
+    /**
+     * First we read in the station data into our local map variable.
+     * @param parameters
+     * @throws Exception
+     */
+    public void open(org.apache.flink.configuration.Configuration parameters) throws Exception {
+        super.open(parameters);
+        //TODO Remove Port after Demo!
```
Comment (skydivin4ng3l) [inline core/src/main/java/org/bptlab/cepta/operators/EnrichDelayWithCoordinatesFunction.java]: remove comment, also next line

File: core/src/main/java/org/bptlab/cepta/operators/EnrichDelayWithCoordinatesFunction.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    @Override
+    public void flatMap(NotificationOuterClass.Notification unenrichedNotification, Collector<NotificationOuterClass.Notification> collector) throws Exception {
+
+        String searchForStationId = unenrichedNotification.getDelay().getStationId().getId();
+        if (coordinateMapping.containsKey(searchForStationId)){
+            CoordinateOuterClass.Coordinate matchingCoordinate = coordinateMapping.get(searchForStationId);
+            NotificationOuterClass.DelayNotification delayNotification = unenrichedNotification.getDelay();
+            //TODO make this less ugly please!
```
Comment (skydivin4ng3l) [inline core/src/main/java/org/bptlab/cepta/operators/EnrichDelayWithCoordinatesFunction.java]: is this already less ugly if so delete this

File: core/src/main/java/org/bptlab/cepta/utils/notification/NotificationHelper.java [hunk 5]
```diff
@@ -5,9 +5,27 @@
 import org.bptlab.cepta.models.internal.delay.DelayOuterClass;
 import org.bptlab.cepta.models.internal.notifications.notification.NotificationOuterClass;
 import org.bptlab.cepta.models.internal.notifications.notification.NotificationOuterClass.Notification;
+import org.bptlab.cepta.models.internal.types.coordinate.CoordinateOuterClass;
 import org.bptlab.cepta.models.internal.types.ids.Ids;
 
 public class NotificationHelper {
+    public static Notification getTrainDelayNotificationFrom(String ceptaID, long delayedSeconds, String delayDetail, long stationId, CoordinateOuterClass.Coordinate coordinate){
```
Comment (skydivin4ng3l) [inline core/src/main/java/org/bptlab/cepta/utils/notification/NotificationHelper.java:12]: maybe change ceptaID -> ceptaIDString?

File: models/events/Stations.proto [hunk 6]
```diff
@@ -0,0 +1,16 @@
+syntax = "proto3";
```
Comment (skydivin4ng3l) [inline models/events/Stations.proto]: Delete this file

File: models/internal/station.proto [hunk 7]
```diff
@@ -3,13 +3,11 @@ package models.internal;
 option go_package = "github.com/bptlab/cepta/models/internal/station";
```
Comment (skydivin4ng3l) [inline models/internal/station.proto]: revert these changes

**Review by skydivin4ng3l — CHANGES_REQUESTED** [review]
please fix

File: models/events/Stations.proto (hunk 6 above)
Comment (Quacck) [reply models/events/Stations.proto]: Done

File: core/src/main/java/org/bptlab/cepta/utils/notification/NotificationHelper.java (hunk 5 above)
Comment (Quacck) [reply core/src/main/java/org/bptlab/cepta/utils/notification/NotificationHelper.java:12]: In its current state the variable is given with the type in front of it so i think thats fine

File: core/src/main/java/org/bptlab/cepta/operators/EnrichDelayWithCoordinatesFunction.java (hunk 4 above)
Comment (Quacck) [reply core/src/main/java/org/bptlab/cepta/operators/EnrichDelayWithCoordinatesFunction.java]: a cool way of  doing this would be to enrich the actual event instead of creating a new one with added information, e.g. sending it to the Builder and adding those fields. But we were not sure how to write that

**Review by skydivin4ng3l — APPROVED** [review]
